networkit/cpp/graph/Graph.cpp

```diff
@@ -828,12 +828,65 @@ void Graph::removeEdgesFromIsolatedSet(const std::vector<node> &nodesInSet) {
 	this->m -= removedEdges;
 }
 
+template <typename Condition>
+count Graph::removeAdjacentEdges(node u, Condition condition, bool edgesIn) {
+	count removedEdges = 0;
+	std::vector<node> &edges_ = outEdges[u];
+	std::vector<edgeweight> &weights_ = edgesIn ? inEdgeWeights[u] : outEdgeWeights[u];
+	std::vector<index> &edgeIds_ = edgesIn ? inEdgeIds[u] : outEdgeIds[u];
+	for (index vi = 0; vi < edges_.size();) {
+		if (condition(edges_[vi])) {
+			++removedEdges;
+			edges_[vi] = edges_.back();
+			edges_.pop_back();
+			if (isWeighted()) {
+				weights_[vi] = weights_.back();
+				weights_.pop_back();
+			}
+			if (hasEdgeIds()) {
+				edgeIds_[vi] = edgeIds_.back();
+				edgeIds_.pop_back();
+			}
+		} else
+			++vi;
+	}
+
+	return removedEdges;
+}
+
 void Graph::removeSelfLoops() {
-	this->forEdges([&](node u, node v, edgeweight) {
-		if (u == v) {
-			removeEdge(u, v);
-		}
+	count nEdgesRemoved = 0;
+
+	forNodes([&](node u) {
+		auto isSelfLoop = [&u](node v) { return u == v; };
+		nEdgesRemoved += removeAdjacentEdges(u, isSelfLoop);
+		if (isDirected())
+			nEdgesRemoved += removeAdjacentEdges(u, isSelfLoop, true);
 	});
+
+	m -= nEdgesRemoved;
```

Comment:
`assert(nEdgesRemoved == numberOfSelfLoops());`?

Reply:
Good point, I added the assertion.